extern/test_bmi_cpp/src/test_bmi_cpp.cpp

```diff
@@ -0,0 +1,429 @@
+#include "test_bmi_cpp.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+
+std::string TestBmiCpp::GetComponentName(){
+  return "Testing BMI C++ Model";
+}
+
+double TestBmiCpp::GetCurrentTime(){
+  return current_model_time;
+}
+
+double TestBmiCpp::GetEndTime(){
+  double time = this->GetStartTime();
+  time += this->num_time_steps * this->time_step_size;
+  return time;
+}
+
+int TestBmiCpp::GetGridRank(const int grid){
+  if (grid == 0) {
+      return 1;
+  }
+  else {
+    throw new std::runtime_error("Rank requested for non-existent grid.");
+  }
+}
+
+int TestBmiCpp::GetGridSize(const int grid){
+  if (grid == 0) {
+      return 1;
+  }
+  else {
+    throw new std::runtime_error("Size requested for non-existent grid.");
+  }
+}
+
+std::string TestBmiCpp::GetGridType(const int grid){
+  if (grid == 0) {
+      return "scalar";
+  }
+  else {
+    throw new std::runtime_error("Type requested for non-existent grid.");
+  }
+}
+
+
+std::vector<std::string> TestBmiCpp::GetInputVarNames(){
+  return input_var_names;  
+}
+std::vector<std::string> TestBmiCpp::GetOutputVarNames(){
+  return output_var_names;
+}
+int TestBmiCpp::GetInputItemCount(){
+  return input_var_names.size();
+}
+int TestBmiCpp::GetOutputItemCount(){
+  return output_var_names.size();
+}
+
+double TestBmiCpp::GetStartTime(){
+  return 0.0;
+}
+
+double TestBmiCpp::GetTimeStep(){
+  return this->time_step_size;
+}
+
+std::string TestBmiCpp::GetTimeUnits(){
+  return "s";
+}
+
+void TestBmiCpp::GetValue(std::string name, void* dest){
+  int inds[] = {0};
+  this->GetValueAtIndices(name, dest, inds, 1);
+}
+
+void TestBmiCpp::GetValueAtIndices(std::string name, void* dest, int* inds, int len){
+  if (len < 1)
+    throw new std::runtime_error("Illegal count provided to GetValueAtIndices(name, dest, inds, count)");
+
+  void *ptr;
+  std::string type;
+  ptr = this->GetValuePtr(name);
+  type = this->GetVarType(name);
+
+  if(type == BMI_TYPE_NAME_DOUBLE){
+    for (size_t i = 0; i < len; ++i) {
+        ((double*)dest)[i] = ((double*)ptr)[inds[i]];
```

Comment:
@hellkite500 Okay, from what I'm reading, yeah, this shouldn't even compile maybe except for some GCC extensions...it's surprising that this compiles on clang in the macOS tests then...but then also, [is this perfectly legal in C](https://github.com/NOAA-OWP/ngen/blob/master/extern/test_bmi_c/src/bmi_test_bmi_c.c#L296)?

Reply:
There is a lot going on in `((double*)dest)[i]`....I think it is legal because it ultimately does produce an lvalue due to `[i]` at the end?  Anyways, in c++ there are more clear semantics that we should probably use.

Reply:
Resolved in 4913934